add chain names to scrapers & output

blackmad opened the pull request.

thismakessand commented:
Some of these spiders scrape multiple brands/chains.  There's an `extras` property already in use on some of these spiders named `brand` that I think should override the `chain_name` property.  Eg: The hilton spider scrapes all the hotel chains in the Hilton family - Waldorf Astoria, Embassy Suites, etc, etc. and the bassproshops spider scrapes Cabela's and Bass Pro Shops locations.  

If there's already a `brand` property in the `extras`, can we move that to `chain_name` and override the class property that is added in this pr?

blackmad commented:
Thanks for pointing that out. Will look & fix.

On Tue, Dec 17, 2019 at 10:50 AM jen w <notifications@github.com> wrote:

> Some of these spiders scrape multiple brands/chains. There's an extras
> property already in use on some of these spiders named brand that I think
> should override the chain_name property. Eg: The hilton spider scrapes
> all the hotel chains in the Hilton family - Waldorf Astoria, Embassy
> Suites, etc, etc. and the bassproshops spider scrapes Cabela's and Bass Pro
> Shops locations.
>
> If there's already a brand property in the extras, can we move that to
> chain_name and override the class property that is added in this pr?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/alltheplaces/alltheplaces/pull/1342?email_source=notifications&email_token=AADMZMDRHVFH7KCN4COHSLLQZDYK5A5CNFSM4J35MV4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHC2IXI#issuecomment-566600797>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADMZMFU5KOIPWDU56VCKT3QZDYK5ANCNFSM4J35MV4A>
> .
>


-- 
*David Blackman*
creative technologist & wandering
help me find my purpose <http://purpose.blackmad.com>

blackmad commented:
It seems maybe slightly cleaner to me to
- rename chain_name to brand
- remove it from the scrapers that scrape multiple brands (I tried to leave
it off the obvious ones, like beermapping)
- only add it to extras if there isn't already a 'brand' field

What do you think?

On Tue, Dec 17, 2019 at 10:51 AM David Blackman <whizziwig@gmail.com> wrote:

> Thanks for pointing that out. Will look & fix.
>
> On Tue, Dec 17, 2019 at 10:50 AM jen w <notifications@github.com> wrote:
>
>> Some of these spiders scrape multiple brands/chains. There's an extras
>> property already in use on some of these spiders named brand that I
>> think should override the chain_name property. Eg: The hilton spider
>> scrapes all the hotel chains in the Hilton family - Waldorf Astoria,
>> Embassy Suites, etc, etc. and the bassproshops spider scrapes Cabela's and
>> Bass Pro Shops locations.
>>
>> If there's already a brand property in the extras, can we move that to
>> chain_name and override the class property that is added in this pr?
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/alltheplaces/alltheplaces/pull/1342?email_source=notifications&email_token=AADMZMDRHVFH7KCN4COHSLLQZDYK5A5CNFSM4J35MV4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHC2IXI#issuecomment-566600797>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AADMZMFU5KOIPWDU56VCKT3QZDYK5ANCNFSM4J35MV4A>
>> .
>>
>
>
> --
> *David Blackman*
> creative technologist & wandering
> help me find my purpose <http://purpose.blackmad.com>
>


-- 
*David Blackman*
creative technologist & wandering
help me find my purpose <http://purpose.blackmad.com>

thismakessand commented:
@blackmad That sounds good to me.

iandees commented on locations/pipelines.py:
```diff
@@ -29,3 +29,13 @@ def process_item(self, item, spider):
+class ApplyBrandPipeline(object):
+
+    def process_item(self, item, spider):
+        existing_extras = item.get('extras', {})
```
I think we should promote `brand` to a documented optional attribute on `DATA_FORMAT.md` and include it in the "item" so that it can be output without having to use the `extras` field. Then this block can check if the brand is on the spider or in the item, preferring the result from the item.

I think we might want to support `brand:wikidata` to let spiders specify the Wikidata QID for the brand.

iandees replied:
I imagine we'll want to include a category at some point, too. Maybe we should model this as a dict of attributes that apply to all items output by a spider, with attributes on the item having precedence?

blackmad replied:
> I think we should promote `brand` to a documented optional attribute on `DATA_FORMAT.md` and include it in the "item" so that it can be output without having to use the `extras` field. Then this block can check if the brand is on the spider or in the item, preferring the result from the item.
> 

are you suggesting that the spiders that already put this in extras/brand, should be edited to move it up to item as well?

bhousel replied:
> I think we might want to support `brand:wikidata` to let spiders specify the Wikidata QID for the brand.

👀 I can help!  want to open a new issue to collaborate on this?

iandees replied:
> are you suggesting that the spiders that already put this in extras/brand, should be edited to move it up to item as well?

Yep!

> I can help! want to open a new issue to collaborate on this?

Hi Bryan! Yea, a new ticket would be great. Thank you!

iandees commented:
This looks great, thanks David!

thismakessand commented:
It doesn't look like this is working as expected.  Previously the Albersons spider produced data with several different brands (Vons, Albertsons, Jewel-Osco, etc), now everything is "Albertsons".  It looks like the brand that is contained in the extras is not overriding the new global `item_attributes` brand.   All the spiders containing `brand` in the `extras` should be updated to move brand to the new `brand` property...

thismakessand commented:
@blackmad ^^

blackmad commented:
Thanks for the bump. Will fix asap. Tried to update them all, not sure why I missed that one. Will figure it out + fix any others

thismakessand commented:
Thanks! I think there's like 20-30 others.  Also there's an unresolved reference in jcrew.py.

blackmad commented:
wow, okay, lots to fix wooooooo.

blackmad commented:
apologies for that.

looked at every scraper that mentioned 'extras' and updated them. fixed jcrew.

https://github.com/alltheplaces/alltheplaces/pull/1351
